architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+considered during the design process.
+-->
+
+#### BicepDeployment Custom Resource Definition
+
+The `BicepDeployment` resource will be a new Kubernetes CRD that will be used to contain the Bicep template and parameters for deploying Radius resources. The CRD will have the following fields:
+
+```go
+// BicepDeploymentSpec defines the desired state of an BicepDeployment
+type BicepDeploymentSpec struct {
+  // Template is the Bicep template that defines the resources to deploy.
+  // +kubebuilder:validation:Required
+  Template string `json:"template"`
+
+  // Parameters is the Bicep parameters file that specifies the parameters for the template.
+  // +kubebuilder:validation:Optional
+  Parameters string `json:"parameters"`
+}
+
+// BicepDeploymentPhrase is a string representation of the current status of an Bicep Deployment.
+type BicepDeploymentPhrase string
+
+const (
+	// BicepDeploymentPhraseUpdating indicates that the Bicep Deployment is being updated.
+	BicepDeploymentPhraseUpdating BicepDeploymentPhrase = "Updating"
+
+	// BicepDeploymentPhraseReady indicates that the Bicep Deployment is ready.
+	BicepDeploymentPhraseReady BicepDeploymentPhrase = "Ready"
+
+	// BicepDeploymentPhraseFailed indicates that the Bicep Deployment has failed.
+	BicepDeploymentPhraseFailed BicepDeploymentPhrase = "Failed"
+
+	// BicepDeploymentPhraseDeleting indicates that the Bicep Deployment is being deleted.
+	BicepDeploymentPhraseDeleting BicepDeploymentPhrase = "Deleting"
+
+	// BicepDeploymentPhraseDeleted indicates that the Bicep Deployment has been deleted.
+	BicepDeploymentPhraseDeleted BicepDeploymentPhrase = "Deleted"
+)
+
+// BicepDeploymentStatus defines the observed state of the Bicep Deployment.
+type BicepDeploymentStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Bicep Deployment. It corresponds to the Deployment's generation, which is updated on mutation by the API Server.
+	// +kubebuilder:validation:Optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
+
+	// Scope is the resource ID of the scope.
+	// +kubebuilder:validation:Optional
+	Scope string `json:"scope,omitempty"`
+
+	// Resource is the resource ID of the deployment.
+	// +kubebuilder:validation:Optional
+	Resource string `json:"resource,omitempty"`
+
+	// Operation tracks the status of an in-progress provisioning operation.
+	// +kubebuilder:validation:Optional
+	Operation *ResourceOperation `json:"operation,omitempty"`
+
+	// Phrase indicates the current status of the Bicep Deployment.
+	// +kubebuilder:validation:Optional
+	Phrase BicepDeploymentPhrase `json:"phrase,omitempty"`
+
+  // Description is a human-readable description of the status of the Bicep Deployment.
+  // +kubebuilder:validation:Optional
+  Description string `json:"description,omitempty"`
+}
+```
+
+#### `rad bicep generate bicepdeployment` Command
```

Comment:
This name is very up for debate.

Reply:
suggestion `rad bicep generate-krm` or `rad bicep generate-kubernetes` or similar.

Additional "words" adds to the command hierarchy. I'm not sure we're going to have additional `rad bicep generate <something else>` commands in the future.